relm-gen-widget/src/lib.rs

```diff
@@ -341,13 +343,62 @@ impl Driver {
+        // This method should probably just be replaced with `impl_view` and
+        // `view_validation_before_impl` should be put inside `impl_view`
+        self.view_validation_before_impl();
+        self.impl_view(name, typ)
+    }
+
+    fn view_validation_before_impl(&mut self) {
+        // This is what comes immediately after `view!` e.g. `{ ... }`
+        let macro_token_tree = &self.view_macro.as_ref().expect("`view!` macro not yet set").tts;
+        // Panic if the macro is declared as anything other than `view! { ... }` or equivalent
+        if macro_token_tree.len() != 1 {
+            panic!("Invalid `view!` syntax, must be `view! { ... }`, `view! ( ... )`, or `view! [ ... ]`");
+        }
+        // Reach inside the brackets and bind the contents (the top level items) of `view!`
+        let top_level_items = match &macro_token_tree[0] {
+            &TokenTree::Delimited(Delimited {ref tts, ..}) => tts.clone(),
+            _ => panic!("Contents of `view!` should be a comma-delimitted series of items")
+        };
+        if top_level_items.len() > 1 {
+            // If the token tree isn't empty, count the number of items at the top level of `view!`
+            let mut comma_previously_found = false;
+            for item in top_level_items {
+                match item {
+                    // Items are separated by commas, so the presence of anything after a comma
+                    // means that there is more than one item at the top level of `view!`. A trailing
+                    // comma should not cause a panic
+                    TokenTree::Token(Token::Comma) => {
+                        if comma_previously_found {
+                            // Bug out if something was found after a comma (another comma in this case)
+                            panic!("There may only be one top-level item in `view!`");
+                        } else {
+                            // Bug out if something was found after a comma
+                            comma_previously_found = true;
+                        }
+                    }
+                    _ => {
+                        if comma_previously_found {
+                            panic!("There may only be one top-level item in `view!`");
+                        }
+                    }
+                }
             }
+        } else if top_level_items.len() == 0 {
+            // Panic if `view!` is empty e.g. `view! {}`
+            panic!("`view!` macro is empty, must contain one top-level item");
+        }
+        let macro_name_segments: &Vec<PathSegment> = &self.view_macro.as_ref().expect("`view!` macro not yet set").path.segments;
+        let last_segment = &macro_name_segments[macro_name_segments.len() - 1];
+        if (macro_name_segments.len() != 1) || (last_segment.ident.as_ref() != "view") {
+            let mut segments: Vec<&str> = Vec::new();
+            for seg in macro_name_segments {
```

Comment:
I guess we could replace this for loop by:
```rust
let joined_path = macro_name_segments.iter()
    .map(|seg| seg.ident.as_ref())
    .collect::<Vec<_>>()
    .join("::");
```

Reply:
I actually did that at first, but couldn't figure out what the "type must be known in this context" error was about. That `<Vec<_>>` is probably what was missing in my code. I'll change that.